xaitk_saliency/impls/gen_detector_prop_sal/drise_scoring.py

```diff
@@ -0,0 +1,132 @@
+    some image, along with predictions for perturbed images and the permutation
+    masks for those images as would be output from a
+    :class:`xaitk_saliency.interfaces.perturb_image.PerturbImage`
+    implementation.
+
+    Object detection representations used here would need to encapsulate
+    localization information (i.e. bounding box regions), class scores, and
+    objectness scores (if applicable to the detector, such as YOLOv3).
+    Object detections are converted into (4+1+nClasses) vectors (4 indices for
+    bounding box locations, 1 index for objectness, and nClasses indices for
+    different object classes).
+    """
+
+    def __init__(
+        self,
+        proximity_metric: str = 'cosine'
+    ):
+
+        try:
+            # Attempting to use chosen comparision metric
+            cdist([[1], [1]], [[1], [1]], proximity_metric)
+            self.proximity_metric: str = proximity_metric
+        except ValueError:
+            raise ValueError("Chosen comparision metric not supported or",
+                             "may not be available in scipy")
+
+    def iou(self, box_a: np.ndarray, box_b: np.ndarray) -> np.ndarray:
+        """Compute the intersection over union (IoU) of two sets of boxes.
+        E.g.:
+            A ∩ B / A ∪ B = A ∩ B / (area(A) + area(B) - A ∩ B)
+        Args:
+            box_a: (np.array) bounding boxes, Shape: [A,4]
+            box_b: (np.array) bounding boxes, Shape: [B,4]
+        Return:
+            iou: (np.array), Shape: [A,B].
+        """
+
+        x11, y11, x12, y12 = np.split(box_a, 4, axis=1)
+        x21, y21, x22, y22 = np.split(box_b, 4, axis=1)
+
+        # determine the (x, y)-coordinates of the intersection rectangle
+        xA = np.maximum(x11, np.transpose(x21))
+        yA = np.maximum(y11, np.transpose(y21))
+        xB = np.minimum(x12, np.transpose(x22))
+        yB = np.minimum(y12, np.transpose(y22))
+
+        # compute the area of intersection rectangle
+        interArea = np.maximum((xB - xA + 1), 0) * np.maximum((yB - yA + 1), 0)
+
+        # compute the area of both the prediction and ground-truth rectangles
+        boxAArea = (x12 - x11 + 1) * (y12 - y11 + 1)
+        boxBArea = (x22 - x21 + 1) * (y22 - y21 + 1)
+
+        # add a small eps to prevent divide by zero errors
+        iou = interArea / \
+            (boxAArea + np.transpose(boxBArea) - interArea + 1e-16)
+
+        return iou
+
+    def generate(
+        self,
+        ref_dets: np.ndarray,
+        perturbed_dets: np.ndarray,
+        perturbed_masks: np.ndarray,
+    ) -> np.ndarray:
+
+        if len(perturbed_dets) != len(perturbed_masks):
+            raise ValueError("Number of perturbation masks and respective",
+                             "detections vector do not match.")
+
+        if ref_dets.shape[1] != perturbed_dets.shape[2]:
+            # TODO: This is not necessarily classes, could also be mismatch
```

Comment:
Essentially, this dimension check (`ref_dets.shape[1]` and `perturbed_dets.shape[2]`) should be `nClasses + 4 + 1`.  However, if there is a shape mismatch it could one of many things: 1) `nClasses` don't agree, 2) one of them is missing the objectness score, or 3) bounding box dimensions are improperly defined. So the `ValueError` of "Number of classes don't match" is not entirely accurate. This was a reminder to change the error message, which I've done so now.